File: Editorial/454_4Sum_II/Java/2_kSum_II.java
import java.util.Map;

class Solution {
    private int[][] lsts;

    public int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
        lsts = new int[][] { A, B, C, D };
        int k = lsts.length;
        Map<Integer, Integer> left = sumCount(0, k / 2);
        Map<Integer, Integer> right = sumCount(k / 2, k);
        int res = 0;
        for (int s : left.keySet())
            res += left.get(s) * right.getOrDefault(-s, 0);
        return res;
    }

    private Map<Integer, Integer> sumCount(int start, int end) {
        Map<Integer, Integer> cnt = new HashMap<>();
        cnt.put(0, 1);
        for (int i = start; i < end; i++) {
            Map<Integer, Integer> map = new HashMap<>();
            for (int a : lsts[i]) {
                for (int total : cnt.keySet()) {
                    map.put(total + a, map.getOrDefault(total + a, 0) + cnt.get(total));
                }
            }
            cnt = map;
        }
        return cnt;
    }
}

File: Editorial/0401-0500/0454_4Sum_II/CPP/2_kSum_II.cpp
class Solution {
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
        vector<vector<int>> lsts = {A, B, C, D};
        int k = int(lsts.size());

        auto countSum = [&](int start, int end) {
            map<int, int> cnt({{0, 1}});
            for (int i = start; i < end; i++) {
                map<int, int> temp;
                for (int total : lsts[i]) {
                    for (auto k = cnt.begin(); k != cnt.end(); k++) {
                        temp[total + k->first] += k->second;
                    }
                }
                cnt = temp;
            }
            return cnt;
        };

        map<int, int> left = countSum(0, k / 2);
        map<int, int> right = countSum(k / 2, k);
        int res = 0;
        for (auto k = left.begin(); k != left.end(); k++) {
            res += k->second * right[-k->first];
        }
        return res;
    }
};
